### ISSUE TITLE
Incorrect inferred title from native callstack

### ISSUE THREAD

**Issue by fmod-mathew** [getsentry/sentry#21433]
## Important Details

How are you running Sentry?
Saas (sentry.io)

## Description

I'm using the store API to write custom callstack based events to Sentry, however I'm not getting the expected title for the issue. I expect the name would be based on the top most stack frame that belongs to my application, however I always get the top most stack frame regardless. I use the custom grouping enhancements to tell Sentry which stack frames belong to my application and which ones don't.

## Steps to Reproduce

1. Using the store endpoint send the following Event:

```json
{
  "platform": "native",
  "level": "fatal",
  "sentry.interfaces.Exception": {
    "values": [
      {
        "type": "Crash",
        "stacktrace": {
          "frames": [
            {
              "function": "Thread_Start",
              "package": "System"
            },
            {
              "function": "Mid_Callstack",
              "package": "MyApp"
            },
            {
              "function": "Crashing_Function",
              "package": "MyApp"
            },
            {
              "function": "DONT_WANT_Exception_Handler",
              "package": "System"
            }
          ]
        }
      }
    ]
  }
}
```

2. Setup a server side custom grouping enhancement like so:
`stack.package:MyApp +app`

3. Check the created issue in Sentry, note the title is "Crash DONT_WANT_Exception_Handler", what I expect is "Crash Crashing_Function".

4. Check the callstack in the issue, Sentry has correctly identified the stack frames from "MyApp" are part of the application and the "System" frames are not.

### Possible Solution

I've done some brief analysis of the Sentry source code, first time so bare with me. It looks like the intended behavior is for the title to be fed via the "culprit" tag. I can see the function `get_culprit` called via `_pull_out_data` from `event_manager.py` takes care of that correctly walking the stack and checking which frames are `in_app`. However the function that applies the server enhancement and ultimately sets `in_app` for the event is called from `normalize_stacktraces_for_grouping`. As far as I can tell these two actions happen around the wrong way, first derive the culprit based on `in_app` then apply the server settings that set `in_app`. I'll leave the solution for someone more familiar with the code, but that's seems to be the core issue.

As for solving my issue, I have the ability to process the stack frames before being sent to Sentry, so I can manually apply the `in_app` tags using my code rather than the server settings. This gives me the desired result, however the above took a bunch of time so I want to leave my notes / findings here for others and potentially for someone to fix.

**Comment by BYK** [comment]
Ping @jan-auer @untitaker

**Comment by jan-auer** [comment]
Thank you for investigating thus far and providing all this context. The logic around generating titles and culprits is indeed complex and probably not documented sufficiently. The primary action item for us is to add missing documentation for this.

To your exception payload next to `type` and `stacktrace`, can you please also add this and then check again:

```json
"mechanism": {
  "type": "generic",
  "synthetic": true
},
```

The reason for this is that we apply different logic depending on the kind of error or crash being reported, and mark ones that are not derived from an actual language-level exception as "synthetic". 

That said, the order of execution here does seem confusing, indeed, and we'll look into this separately.

**Comment by fmod-mathew** [comment]
Thanks, that actually fixes the problem for me, the issue name is assigned based on the top in_app frame after the server side grouping enhancements now.

In fact I added the mechanism type and synthetic to my payload recently to fix another issue (issue names in discover weren't correct). I was concerned by using the synthetic tag because it's not documented, I figured it out by comparing the JSON from one of our minidump issues compared to the manual payload I'm sending.

Improvements to the documentation surrounding this would be great, especially documenting the importance of the synthetic tag. Thanks again for your support.
